extensions/ql-vscode/src/run-remote-query.ts

```diff
@@ -1,20 +1,51 @@
-import { Uri } from 'vscode';
+import { Uri, window } from 'vscode';
 import * as yaml from 'js-yaml';
 import * as fs from 'fs-extra';
-import { showAndLogErrorMessage, showAndLogInformationMessage } from './helpers';
+import { getOnDiskWorkspaceFolders, showAndLogErrorMessage, showAndLogInformationMessage } from './helpers';
 import { Credentials } from './authentication';
+import * as cli from './cli';
+import { logger } from './logging';
 
 interface Config {
   repositories: string[];
   ref?: string;
-  language: string;
+  language?: string;
 }
 
 // Test "controller" repository and workflow.
 const OWNER = 'dsp-testing';
 const REPO = 'qc-controller';
 
-export default async function runRemoteQuery(credentials: Credentials, uri?: Uri) {
+/**
+ * Finds the language that a query targets.
+ * If it can't be autodetected, prompt the user to specify the language manually.
+ */
+export async function findLanguage(
```

Comment:
Since you asked about tests: I agree that having tests for this function would be handy, and give you confidence in its behaviour. However we don't yet have a test module for `run-remote-query` due to the experimental nature of the entire feature, so adding a test requires some upfront work. (In future it will be easier :) )

I'll record some pointers here about how you could create those tests, but I'm not expecting you to do it in this PR. We can look at test coverage together after this PR.
- For integration tests (which use a real query and actually run the underlying commands), copy `extensions/ql-vscode/src/vscode-tests/cli-integration/queries.test.ts` as a template. I think it would make sense to have:
  - one test function that runs `findLanguage` on a query with a real language dependency and identifies the language
  - one test function that runs it on a dummy query and returns `undefined`
  - any other cases you can think of
- For unit tests (which don't run the actual commands, but simulate the underlying behaviour), use `extensions/ql-vscode/src/vscode-tests/no-workspace/query-results.test.ts` as a template. Those tests mock the behaviour of the CLI server.